wenet/transducer/transducer.py

```diff
@@ -0,0 +1,334 @@
+                normalize_length=length_normalized_loss,
+            )
+
+    def forward(
+        self,
+        speech: torch.Tensor,
+        speech_lengths: torch.Tensor,
+        text: torch.Tensor,
+        text_lengths: torch.Tensor,
+    ) -> Dict[str, Optional[torch.Tensor]]:
+        """Frontend + Encoder + predictor + joint + loss
+
+        Args:
+            speech: (Batch, Length, ...)
+            speech_lengths: (Batch, )
+            text: (Batch, Length)
+            text_lengths: (Batch,)
+        """
+        assert text_lengths.dim() == 1, text_lengths.shape
+        # Check that batch_size is unified
+        assert (speech.shape[0] == speech_lengths.shape[0] == text.shape[0] ==
+                text_lengths.shape[0]), (speech.shape, speech_lengths.shape,
+                                         text.shape, text_lengths.shape)
+
+        # Encoder
+        encoder_out, encoder_mask = self.encoder(speech, speech_lengths)
+        encoder_out_lens = encoder_mask.squeeze(1).sum(1)
+        # predictor
+        ys_in_pad = add_blank(text, self.blank)
+        predictor_out = self.predictor(ys_in_pad)
+        # joint
+        joint_out = self.joint(encoder_out, predictor_out)
+        # NOTE(Mddct): some loss implementation require pad valid is zero
+        # torch.int32 rnnt_loss required
+        text = text.to(torch.int64)
+        text = torch.where(text == self.ignore_id, 0, text).to(torch.int32)
+        text_lengths = text_lengths.to(torch.int32)
+        encoder_out_lens = encoder_out_lens.to(torch.int32)
+        loss = torchaudio.functional.rnnt_loss(joint_out,
+                                               text,
+                                               encoder_out_lens,
+                                               text_lengths,
+                                               blank=self.blank,
+                                               reduction="mean")
+        loss_rnnt = loss
+
+        # optional attention decoder
+        loss_att: Optional[torch.Tensor] = None
+        if self.transducer_weight != 1.0 and self.attention_decoder is not None:
+            loss_att, _ = self._calc_att_loss(encoder_out, encoder_mask, text,
+                                              text_lengths)
+
+        # optional ctc
+        loss_ctc: Optional[torch.Tensor] = None
+        if self.ctc_weight != 0.0 and self.ctc is not None:
+            loss_ctc = self.ctc(encoder_out, encoder_out_lens, text,
+                                text_lengths)
+        else:
+            loss_ctc = None
+
+        if loss_ctc is not None:
+            loss = loss + self.ctc_weight * loss_ctc.sum()
+        if loss_att is not None:
+            loss = loss + self.attention_decoder_weight * loss_att.sum()
+        # NOTE: 'loss' must be in dict
+        return {
+            'loss': loss,
+            'loss_rnnt': loss_rnnt,
+            'loss_att': loss_att,
+            'loss_ctc': loss_ctc
+        }
```

Comment:
It's better that we keep the order compatiable with joint CTC/attention model. namely `loss`, `loss_att`, loss_ctc`, `loss_rnnt`, it's also in alphabet order.